crates/ide-assists/src/handlers/bool_to_enum.rs

```diff
@@ -0,0 +1,724 @@
+        }
+        Some(BoolNodeData {
+            target_node: strukt.syntax().clone(),
+            name: field_name,
+            ty_annotation: field.ty(),
+            initializer: None,
+            definition: Definition::Field(def),
+        })
+    } else {
+        None
+    }
+}
+
+fn replace_bool_expr(edit: &mut SourceChangeBuilder, expr: ast::Expr) {
+    let expr_range = expr.syntax().text_range();
+    let enum_expr = bool_expr_to_enum_expr(expr);
+    edit.replace(expr_range, enum_expr.syntax().text())
+}
+
+/// Converts an expression of type `bool` to one of the new enum type.
+fn bool_expr_to_enum_expr(expr: ast::Expr) -> ast::Expr {
+    let true_expr = make::expr_path(make::path_from_text("Bool::True")).clone_for_update();
+    let false_expr = make::expr_path(make::path_from_text("Bool::False")).clone_for_update();
+
+    if let ast::Expr::Literal(literal) = &expr {
+        match literal.kind() {
+            ast::LiteralKind::Bool(true) => true_expr,
+            ast::LiteralKind::Bool(false) => false_expr,
+            _ => expr,
+        }
+    } else {
+        make::expr_if(
+            expr,
+            make::tail_only_block_expr(true_expr),
+            Some(ast::ElseBranch::Block(make::tail_only_block_expr(false_expr))),
+        )
+        .clone_for_update()
+    }
+}
+
+/// Replaces all usages of the target identifier, both when read and written to.
+fn replace_usages(edit: &mut SourceChangeBuilder, usages: &UsageSearchResult) {
+    for (_, references) in usages.iter() {
+        references
+            .into_iter()
+            .filter_map(|FileReference { range, name, .. }| match name {
+                ast::NameLike::NameRef(name) => Some((*range, name)),
+                _ => None,
+            })
+            .for_each(|(range, name_ref)| {
+                if let Some(initializer) = find_assignment_usage(name_ref) {
+                    cov_mark::hit!(replaces_assignment);
+
+                    replace_bool_expr(edit, initializer);
+                } else if let Some((prefix_expr, expr)) = find_negated_usage(name_ref) {
+                    cov_mark::hit!(replaces_negation);
+
+                    edit.replace(
+                        prefix_expr.syntax().text_range(),
+                        format!("{} == Bool::False", expr),
+                    );
+                } else if let Some((record_field, initializer)) = find_record_expr_usage(name_ref) {
+                    cov_mark::hit!(replaces_record_expr);
+
+                    let record_field = edit.make_mut(record_field);
+                    let enum_expr = bool_expr_to_enum_expr(initializer);
+                    record_field.replace_expr(enum_expr);
+                } else if name_ref.syntax().ancestors().find_map(ast::Expr::cast).is_some() {
+                    // for any other usage in an expression, replace it with a check that it is the true variant
+                    edit.replace(range, format!("{} == Bool::True", name_ref.text()));
```

Comment:
Is there a convention for when to use `edit.replace` vs `ted::replace`? I think it seems nicer to use `ted` instead of a formatted string but I got a bit tripped up with when to make things mutable

Reply:
There isn't really a single proper way to editing unfortunately, either should be fine

Reply:
Ah okay, no problem then thanks!